Design.md

```diff
@@ -0,0 +1,107 @@
+
+Proposed key generation scheme:
+* <genkey> := <version>#<topology_version>#<split>#<rep_policy>#<object_key>#<rand>#<location>[#<location>]
+* <version> := Natural (so 0 or 1 to start)
+* <topology_version> := Natural (so 0 or 1 to start) - if we use indirection instead of storing raw UUIDS we must versioned the mapping
+* <split> := <n_chunk>,<split_size>
+* <n_chunk> := number of splitted parts (1 for non-splitted objects)
+* <split_size> := size of each splitted parts, except last one (see hyperdrive keys below)
+* <rep_policy> := RS,k,m or CP,n (for n-1 copies)
+* <object_key> := parent S3 object key (or a prefix) - can contain anything but ‘#’
+* <rand> := 64 bits random number (unicity inside 1 hyperdrive)
+* <location>:= hyperdrive location (UUID, idx in table?, ip:port, ...)
+
+The fragments do not store the exact, really used hyperdrive keys. Those can be derived easily with only the generated key, even for split.
+* <stored_fragment_key> := <object_key>-<rand>-<start_offset>-<type>-<fragid>
+* <object_key> and <rand> are the ones defined above
+* <type> := ‘d’ for data, ‘c’ for coding
+* <fragid>:= index in main key fragment list
+* <start_offset> := used for splits. All split chunks share the same prefix, storing the offset is used to easily have range queries and avoid storing them all in the main key.
+
+Meets all requirements (except perhaps around the last stripe size of an ECN chunk). Split is required to contact the same hyperdrives for the object. This has several benefits: only selecting once, object located on only a handful of hyperdrives => less machines to contact = less opportunities to fail or hit a straggler. The way split is handled also enables us not to have a manifest, worries about its size, location or freshness.
+
+Example key:
+1/ Small key: storing s3://fakebucket/obj1 with RS2+1 onto hd1 hd2 and hd3
+Main key: 1#1#1,0#RS,2,1#obj1#314159#hd1#hd3#hd2
+Hyperdrive keys:
+1. obj1-314159-0-d-1
+2. obj1-314159-0-d-2
+3. obj1-314159-0-c-3
+
+2/ Splitted key: storing s3://fakebucket/Large1 with RS2+1 onto hd1 hd2 and hd3
+Key size: 64000, split_size: 49000, 2 parts on same hyperdrive (no conflict!)
+
+Main key: 1#1#2,49000#RS,2,1#Large1#42#hd3#hd2#hd3
+Hyperdrive keys:
+* On hd1: none
+* On hd2
+  1. Large1-42-0-d-2
+  2. Large1-42-49000-d-2
+* On hd3
+  1. Large1-42-0-d-1
+  2. Large1-42-0-c-3
+  3. Large1-42-49000-d-1
+  4. Large1-42-49000-c-3
+
+## Error handling
+
+### On DELETE
+
+Failure to delete a fragment ( != 20X, 404) should not fail the overall operation but must log an appropriate entry into designated Kafka topic, to be cleaned later on by HdRepair.
+
+Topic: orphans
+Content: JSON messages
+Example JSON:
+```json
+{
+    "objectKey": ... # S3 object key
+    "rawKey": .... # associated hdclient metadata
+    "orphans": [
+        hyperdrive_key_missing_fragment1,
```

Comment:
In the case of an orphan, it would not be a missing fragment, but instead a remaining and not-cleaned-up fragment, right ? (would make things clearer/more intuitive to me at least)

Reply:
Very much so.